espflash/src/cli/monitor/stack_dump.rs

```diff
@@ -0,0 +1,204 @@
+
+use gimli::{Section, UnwindSection};
+use object::{Object, ObjectSection};
+
+use crate::cli::monitor::symbols::Symbols;
+
+pub(crate) const MARKER: &str = "STACKDUMP: ";
+
+pub(crate) fn backtrace_from_stack_dump(
+    line: &str,
+    out: &mut dyn Write,
+    elf: Option<&[u8]>,
+    symbols: &Symbols<'_>,
+) -> std::io::Result<()> {
+    if let Some(elf) = elf {
+        if let Some(remaining) = line.to_string().strip_prefix(MARKER) {
+            let mut split = remaining.split(" ");
+            let (address, stack) = {
+                let first = split.next();
+                let second = split.next();
+
+                (first, second)
+            };
+
+            if let Some(address) = address {
+                if let Some(stack) = stack {
+                    if stack.len() % 2 != 0 {
+                        return Ok(());
+                    }
+
+                    let mut pc = u32::from_str_radix(address, 16).unwrap_or_default();
+                    let mut stack_bytes = Vec::new();
+                    for byte_chars in stack.chars().collect::<Vec<char>>().chunks(2) {
+                        if byte_chars.len() == 2 {
+                            stack_bytes.push(
+                                u8::from_str_radix(
+                                    &format!("{}{}", byte_chars[0], byte_chars[1]),
+                                    16,
+                                )
+                                .unwrap_or_default(),
+                            );
+                        }
+                    }
+
+                    let func_info = get_func_info(elf)?;
+
+                    writeln!(out).ok();
+                    let mut index = 0;
+                    loop {
+                        let func = func_info.iter().find(|f| f.start <= pc && f.end >= pc);
+                        if let Some(func) = func {
+                            if func.stack_frame_size == 0 {
+                                break;
+                            }
+
+                            let lookup_pc = pc as u64 - 4;
+                            let name = symbols.name(lookup_pc);
+                            let location = symbols.location(lookup_pc);
+                            if let Some(name) = name {
+                                if let Some((file, line_num)) = location {
+                                    writeln!(out, "{name}\r\n    at {file}:{line_num}\r\n").ok();
+                                } else {
+                                    writeln!(out, "{name}\r\n    at ??:??\r\n").ok();
+                                }
+                            }
+
+                            if index + func.stack_frame_size as usize > stack_bytes.len() {
+                                break;
+                            }
+
+                            let next_pc_pos = index + (func.stack_frame_size as usize - 4);
+
+                            pc = u32::from_le_bytes(
```

Comment:
Won't this blow up in trap handlers? MEPC is not on the stack, at least not in a stable position, so we can't unwind past a trap handler, can we? Likely we should look out for `.cfi_underfined ra` markers, they show up as RegisterRule::Undefined in the debuginfo.